libvast/src/segment_store.cpp

```diff
@@ -177,27 +170,135 @@ std::unique_ptr<store::lookup> segment_store::extract(const ids& xs) const {
+    return contains(xs, ys);
+  };
+  // Counts number of total erased events for user-facing output.
+  uint64_t erased_events = 0;
+  // Convenience function for updating given segment by pruning all IDs in `xs`
+  // from it.
+  // @returns `true` if the entire segment got erased, `false` otherwise.
+  auto update = [&](auto& sref) {
+    auto slice_ids = sref.meta().slice_ids();
+    // Check whether we can drop the entire segment.
+    if (std::all_of(slice_ids.begin(), slice_ids.end(), is_subset_of_xs)) {
+      erased_events += drop(sref);
+      return;
+    }
+    auto segment_id = sref.id();
+    // Get all slices in the segment and generate a new segment that contains
+    // only what's left after dropping the selection.
+    auto segment_ids = sref.meta().flat_slice_ids();
+    std::vector<table_slice_ptr> slices;
+    if (auto maybe_slices = sref.lookup(segment_ids)) {
+      slices = std::move(*maybe_slices);
+      if (slices.empty()) {
+        VAST_WARNING(this, "got no slices after lookup for segment", segment_id,
+                     "=> erases entire segment!");
+        erased_events += drop(sref);
+        return;
+      }
+    } else {
+      VAST_WARNING(this, "was unable to get table slice for segment",
+                   segment_id, "=> erases entire segment!");
+      erased_events += drop(sref);
+      return;
+    }
+    VAST_ASSERT(slices.size() > 0);
+    auto keep_mask = ~xs;
+    std::vector<table_slice_ptr> new_slices;
+    for (auto& slice : slices) {
+      // Expand keep_mask on-the-fly if needed.
+      auto max_id = slice->offset() + slice->rows();
+      if (keep_mask.size() < max_id)
+        keep_mask.append_bits(true, max_id - keep_mask.size());
+      size_t new_slices_size_before = new_slices.size();
+      select(new_slices, slice, keep_mask);
+      size_t remaining_rows = 0;
+      for (size_t i = new_slices_size_before; i < new_slices.size(); ++i)
+        remaining_rows += new_slices[i]->rows();
+      erased_events += slice->rows() - remaining_rows;
+    }
+    if (new_slices.empty()) {
+      VAST_WARNING(this, "was unable to generate any new slice for segment",
+                   segment_id, "=> erases entire segment!");
+      erased_events += drop(sref);
+      return;
+    }
+    VAST_DEBUG(this, "shrinks segment", segment_id, "from", slices.size(), "to",
+               new_slices.size(), "slices");
+    // Remove stale state.
+    segments_.erase_value(segment_id);
+    // Create a new segment from the remaining slices.
+    segment_builder tmp_builder;
+    segment_builder* builder = &tmp_builder;
+    if constexpr (std::is_same_v<decltype(sref), segment_builder&>) {
+      // If `update` got called with a builder then we simply use that by
+      // resetting it and filling it with new content. Otherwise, we fill
+      // `tmp_builder` instead and replace the the segment `sref` in the next
+      // `if constexpr` block.
```

Comment:
This is very useful comment. Can you add it to the definition of the function? It would be helpful there to know that the lambda is actually generic.